README.md
# Logstash Input for CloudWatch Logs

[![Gem][ico-version]][link-rubygems]
[![Downloads][ico-downloads]][link-rubygems]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]

> Stream events from CloudWatch Logs.

### Purpose
Specify an individual log group or array of groups, and this plugin will scan
all log streams in that group, and pull in any new log events.

Optionally, you may set the `log_group_prefix` parameter to true
which will scan for all log groups matching the specified prefix(s)
and ingest all logs available in all of the matching groups.

The optional `log_group_suffix` parameter will use the results of the above
cloudwatch query and then filter them

## Usage

### Parameters
| Parameter | Input Type | Required | Default |
|-----------|------------|----------|---------|
| log_group | string or Array of strings | Yes | |
| log_group_prefix | boolean | No | `false` |
| log_group_suffix | string | No | |
| start_position | `beginning`, `end`, or an Integer | No | `beginning` |
| sincedb_path | string | No | `$HOME/.sincedb*` |
| interval | number | No | 60 |
| aws_credentials_file | string | No | |
| access_key_id | string | No | |
| secret_access_key | string | No | |
| session_token | string | No | |
| region | string | No | `us-east-1` |
| codec | string | No | `plain` |

#### `start_position`
The `start_position` setting allows you to specify where to begin processing
a newly encountered log group on plugin boot. Whether the group is 'new' is
determined by whether or not the log group has a previously existing entry in
the sincedb file.

Valid options for `start_position` are:
* `beginning` - Reads from the beginning of the group (default)
* `end` - Sets the sincedb to now, and reads any new messages going forward
* Integer - Number of seconds in the past to begin reading at

#### Logstash Default config params
Other standard logstash parameters are available such as:
* `add_field`
* `type`
* `tags`

### Example

    input {
        cloudwatch_logs {
            log_group => [ "/aws/lambda/my-lambda" ]
            access_key_id => "AKIAXXXXXX" 
            secret_access_key => "SECRET"
        }
    }

## Development
The [default logstash README](DEVELOPER.md) which contains development directions and other information has been moved to [DEVELOPER.md](DEVELOPER.md).

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elasticsearch/logstash/blob/master/CONTRIBUTING.md) file.

[ico-version]: https://img.shields.io/gem/v/logstash-input-cloudwatch_logs.svg?style=flat-square
[ico-downloads]: https://img.shields.io/gem/dt/logstash-input-cloudwatch_logs.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/License-Apache%202.0-blue.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/lukewaite/logstash-input-cloudwatch-logs.svg?style=flat-square

[link-rubygems]: https://rubygems.org/gems/logstash-input-cloudwatch_logs
[link-travis]: https://travis-ci.org/lukewaite/logstash-input-cloudwatch_logs


CHANGELOG.md
# Release Notes for v1.x

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

* Updated sincedb to track the high water mark timestamps at a stream level, rather than just at the group level. This is a fix for [#74](https://github.com/lukewaite/logstash-input-cloudwatch-logs/issues/74)

## [v1.0.3] (2018-05-28)
* Update minimum version of `logstash-mixin-aws` to support assume role ([#9a4677f](https://github.com/lukewaite/logstash-input-cloudwatch-logs/commit/9a4677fef8bcbf291bd4b357be2a9568ea4f3fc1) - Fixes [#51](https://github.com/lukewaite/logstash-input-cloudwatch-logs/issues/51), [#39](https://github.com/lukewaite/logstash-input-cloudwatch-logs/issues/39))

## [v1.0.2] (2017-10-07)
* Handle a `start_position` integer value correctly ([#41](https://github.com/lukewaite/logstash-input-cloudwatch-logs/pull/41) - Fixes [#38](https://github.com/lukewaite/logstash-input-cloudwatch-logs/issues/38))

## [v1.0.1] (2017-08-22)

### Fixed
* Handle rate limiting when making requests to the CWL apis ([#32](https://github.com/lukewaite/logstash-input-cloudwatch-logs/pull/32))

## [v1.0.0] (2017-07-15)

### Added
* Implement `start_position` setting ([#28](https://github.com/lukewaite/logstash-input-cloudwatch-logs/issues/28))
* Allow log_group to be an array of groups (or prefixes if enabled) ([#13](https://github.com/lukewaite/logstash-input-cloudwatch-logs/issues/13))

### Fixed
* Ensure the plugin stops properly
* Relax the contstraint on `logstash-mixin-aws` supporting Logstash 2.4
* Refactored ingestion, fixes multiple memory leaks (Fixes [#24](https://github.com/lukewaite/logstash-input-cloudwatch-logs/issues/4))
* Pull only log_events since last ingestion (Fixes [#10](https://github.com/lukewaite/logstash-input-cloudwatch-logs/issues/10))
* Incrementally write to since_db on each page of data from the CWL API (Fixes [#4](https://github.com/lukewaite/logstash-input-cloudwatch-logs/issues/4))

[Unreleased]: https://github.com/lukewaite/logstash-input-cloudwatch-logs/compare/v1.0.3...HEAD
[v1.0.3]: https://github.com/lukewaite/logstash-input-cloudwatch-logs/compare/v1.0.2...v1.0.3
[v1.0.2]: https://github.com/lukewaite/logstash-input-cloudwatch-logs/compare/v1.0.1...v1.0.2
[v1.0.1]: https://github.com/lukewaite/logstash-input-cloudwatch-logs/compare/v1.0.0...v1.0.1
[v1.0.0]: https://github.com/lukewaite/logstash-input-cloudwatch-logs/compare/v0.10.3...v1.0.0


spec/inputs/cloudwatch_logs_spec.rb
# encoding: utf-8
require 'logstash/devutils/rspec/spec_helper'
require 'logstash/inputs/cloudwatch_logs'
require 'aws-sdk-resources'
require 'aws-sdk'
require "logstash/timestamp"

describe LogStash::Inputs::CloudWatch_Logs do
  def parse_time(data)
    LogStash::Timestamp.at(data.to_i / 1000, (data.to_i % 1000) * 1000)
  end # def parse_time


  let(:config) {
    {
        'access_key_id' => '1234',
        'secret_access_key' => 'secret',
        'log_group' => ['sample-log-group'],
        'region' => 'us-east-1'
    }
  }

  before do
    Aws.config[:stub_responses] = true
    Thread.abort_on_exception = true
  end


  context 'when interrupting the plugin' do
    let(:config) {super.merge({'interval' => 5})}

    before do
      expect_any_instance_of(LogStash::Inputs::CloudWatch_Logs).to receive(:process_group).and_return(nil)
    end

    it_behaves_like 'an interruptible input plugin'
  end

  describe '#register' do
    context 'default config' do
      subject {LogStash::Inputs::CloudWatch_Logs.new(config)}

      it 'registers succesfully' do
        expect {subject.register}.to_not raise_error
      end
    end

    context 'start_position set to end' do
      subject {LogStash::Inputs::CloudWatch_Logs.new(config.merge({'start_position' => 'end'}))}

      it 'registers succesfully' do
        expect {subject.register}.to_not raise_error
      end
    end

    context 'start_position set to an integer' do
      subject {LogStash::Inputs::CloudWatch_Logs.new(config.merge({'start_position' => 100}))}

      it 'registers succesfully' do
        expect {subject.register}.to_not raise_error
      end
    end

    context 'start_position invalid' do
      subject {LogStash::Inputs::CloudWatch_Logs.new(config.merge({'start_position' => 'invalid start position'}))}

      it 'raises a configuration error' do
        expect {subject.register}.to raise_error(LogStash::ConfigurationError)
      end
    end
  end


  describe '#find_log_groups without prefix true' do
    context 'with an array in the config' do
      subject {LogStash::Inputs::CloudWatch_Logs.new(config)}

      it 'passes through configuration' do
        expect(subject.find_log_groups).to eq(['sample-log-group'])
      end
    end

    context 'with a single string in the log_group' do
      subject {LogStash::Inputs::CloudWatch_Logs.new(config.merge({'log_group' => 'sample-log-group-string'}))}

      it 'array-ifies the single string' do
        expect(subject.find_log_groups).to eq(['sample-log-group-string'])
      end
    end
  end

  
  describe '#process_log' do
    context 'with an array in the config' do      
      subject {LogStash::Inputs::CloudWatch_Logs.new(config.merge({
        'start_position' => 0,
        'sincedb_path' => Dir.mktmpdir("rspec-")  + '/sincedb.txt'
        }))}

      it 'check  default start time - beginning ' do
        # given the config is for beginning
        subject.instance_variable_set(:@start_position, 'beginning')

        # then the default time is epoch start 
        expect(subject.send(:get_default_start_time, *[])).to eq(0)
      end

      it 'check  default start time - end ' do
        # given the config is for end
        subject.instance_variable_set(:@start_position, 'end')

        # then the default time nearly now 
        now = DateTime.now.strftime('%Q').to_i
        expect(subject.send(:get_default_start_time, *[])).to be_within(100).of(now)
      end

      it 'check  default start time - start position ' do
        # given the config is for end
        subject.instance_variable_set(:@start_position, '86400')

        # then the default time nearly the expected  
        now = DateTime.now.strftime('%Q').to_i
        expected = now - 86400 * 1000
        expect(subject.send(:get_default_start_time, *[])).to be_within(100).of(expected)
      end      

      it 'process a log event - event is new' do
        subject.register
        event_tracker = subject.instance_variable_get(:@event_tracker)

        # given these times
        old_timestamp = DateTime.now.strftime('%Q').to_i
        new_timestamp = old_timestamp + 1000

        # given we know about this group and stream from an old record
        group = 'groupA'      
        # given we got the message for this group, for the known stream
        # and where the record is "new enough"
        log = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
        log.message = 'this be the verse'
        log.timestamp = old_timestamp
        log.ingestion_time = 123
        log.log_stream_name = 'streamX'
        log.event_id = 'event1'

        event_tracker.record_processed_event(group, log)

        # update this log to be a new event
        log.timestamp = new_timestamp
        log.event_id = 'event2'

        # when we send the log (assuming we have a queue)
        queue = []
        subject.instance_variable_set(:@queue, queue)

        subject.send(:process_log, *[log, group])

        # then a message was sent to the queue
        expect(queue.length).to eq(1)

        
        expect(queue[0].get('[@timestamp]')).to eq(LogStash::Timestamp.at(new_timestamp.to_i / 1000, (new_timestamp.to_i % 1000) * 1000))
        expect(queue[0].get('[message]')).to eq('this be the verse')
        expect(queue[0].get('[cloudwatch_logs][log_group]')).to eq('groupA')
        expect(queue[0].get('[cloudwatch_logs][log_stream]')).to eq('streamX')
        expect(queue[0].get('[cloudwatch_logs][ingestion_time]').to_iso8601).to eq('1970-01-01T00:00:00.123Z')

        # then the timestamp should have been updated
        start_time = event_tracker.min_time(group)
        # and the new start time the earliest record
        expect(start_time).to eq(old_timestamp)
      end
    
    

      it 'process a log event - event is old' do
        subject.register
        event_tracker = subject.instance_variable_get(:@event_tracker)

        # given these times
        old_timestamp = DateTime.now.strftime('%Q').to_i
        new_timestamp = old_timestamp + 1000


        puts("old_timestamp: #{old_timestamp}")
        puts("new_timestamp: #{new_timestamp}")


        # given we previously got the old record
        group = 'GroupA'
        log = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
        log.message = 'this be the verse'
        log.timestamp = new_timestamp
        log.ingestion_time = 123
        log.log_stream_name = 'streamX'
        log.event_id = 'eventA'
        event_tracker.record_processed_event(group, log)

        # given a new log message
        log.timestamp = old_timestamp
        log.event_id = 'eventB'


        # when we send the log (assuming we have a queue)
        queue = []
        subject.instance_variable_set(:@queue, queue)

        subject.send(:process_log, *[log, group])

        # then no message was sent to the queue
        expect(queue.length).to eq(0)

        # then the timestamp should not have been updated
        start_time = event_tracker.min_time(group)
        # and the new start time is 1 millisecond after the message time
        expect(start_time).to eq(new_timestamp)

      end    

      it 'process a log event - event has already been seen' do
        subject.register
        event_tracker = subject.instance_variable_get(:@event_tracker)

        # given these times
        old_timestamp = DateTime.now.strftime('%Q').to_i


        # given we previously got the old record
        group = 'GroupA'
        log = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
        log.message = 'this be the verse'
        log.timestamp = old_timestamp
        log.ingestion_time = 123
        log.log_stream_name = 'streamX'
        log.event_id = 'eventA'
        event_tracker.record_processed_event(group, log)


        # when we send the log (assuming we have a queue)
        queue = []
        subject.instance_variable_set(:@queue, queue)

        subject.send(:process_log, *[log, group])

        # then no message was sent to the queue
        expect(queue.length).to eq(0)

        # then the timestamp should not have been updated
        start_time = event_tracker.min_time(group)
        # and the new start time is 1 millisecond after the message time
        expect(start_time).to eq(old_timestamp)

      end          
    end

    
  end    

end


spec/inputs/last_event_tracker_spec.rb
# encoding: utf-8
require 'logstash/devutils/rspec/spec_helper'
require 'logstash/inputs/last_event_tracker'
require 'aws-sdk-resources'
require 'aws-sdk'

describe LastEventTracker do

  describe "test add events" do
    
    it "add single event"   do
      tracker = LastEventTracker.new(Dir.mktmpdir("rspec-")  + '/sincedb.txt', 15)

      # given a new log event
      log = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
      log.message = 'this be the verse'
      log.timestamp = 1
      log.log_stream_name = 'streamX'
      log.event_id = 'event1'

      # when we push the first event, then it will not have been processed
      expect(tracker.check_event_already_processed('group', log)).to be_falsey
      
      # and it will be added 
      expect(tracker.get('group', log).events).to contain_exactly(log.event_id)
      expect(tracker.get('group', log).timestamp).to eq(log.timestamp)                 
    end


    
    it "add multiple events in the same time period"   do
      tracker = LastEventTracker.new(Dir.mktmpdir("rspec-")  + '/sincedb.txt', 15)

      # given multiple log events in the same time period
      log1 = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
      log1.timestamp = 1
      log1.log_stream_name = 'streamX'
      log1.event_id = 'event1'

      log2 = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
      log2.timestamp = log1.timestamp
      log2.log_stream_name = 'streamX'
      log2.event_id = 'event2'
      
      log3 = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
      log3.timestamp = log1.timestamp
      log3.log_stream_name = 'streamX'
      log3.event_id = 'event3'      

      # when we push the first event, then it will not have been processed
      expect(tracker.check_event_already_processed('group', log1)).to be_falsey
      expect(tracker.check_event_already_processed('group', log2)).to be_falsey
      expect(tracker.check_event_already_processed('group', log3)).to be_falsey
      
      # and they will be added 
      expect(tracker.get('group', log3).events).to contain_exactly(log1.event_id, log2.event_id, log3.event_id)
      expect(tracker.get('group', log3).timestamp).to eq(log1.timestamp)                 
    end

    it "add duplicate event"   do
      tracker = LastEventTracker.new(Dir.mktmpdir("rspec-")  + '/sincedb.txt', 15)

      # given a new log event
      log = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
      log.message = 'this be the verse'
      log.timestamp = 1
      log.log_stream_name = 'streamX'
      log.event_id = 'event1'

      # when we push the first event, then it will not have been processed
      expect(tracker.check_event_already_processed('group', log)).to be_falsey
      expect(tracker.check_event_already_processed('group', log)).to be_truthy
      
      # and it will be added 
      expect(tracker.get('group', log).events).to contain_exactly(log.event_id)
      expect(tracker.get('group', log).timestamp).to eq(log.timestamp)    
      
      tracker.save()
    end


    it "add multiple events in a new time period"   do
      tracker = LastEventTracker.new(Dir.mktmpdir("rspec-")  + '/sincedb.txt', 15)

      # given multiple log events in the same time period
      log1 = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
      log1.timestamp = 1
      log1.log_stream_name = 'streamX'
      log1.event_id = 'event1'

      log2 = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
      log2.timestamp = log1.timestamp
      log2.log_stream_name = 'streamX'
      log2.event_id = 'event2'
      
      log3 = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
      log3.timestamp = log1.timestamp + 1
      log3.log_stream_name = 'streamX'
      log3.event_id = 'event3'      

      # when we push the first event, then it will not have been processed
      expect(tracker.check_event_already_processed('group', log1)).to be_falsey
      expect(tracker.check_event_already_processed('group', log2)).to be_falsey
      expect(tracker.check_event_already_processed('group', log3)).to be_falsey
      
      # and they will be added 
      expect(tracker.get('group', log3).events).to contain_exactly(log3.event_id)
      expect(tracker.get('group', log3).timestamp).to eq(log3.timestamp)                 
    end


    it "test file handling"   do
      pathToFile = Dir.mktmpdir("rspec-")  + '/sincedb.txt'

      puts ("Testing with #{pathToFile}")
      File.open(pathToFile, 'w') { |file| file.write('{"group.streamX": {"timestamp": 1, "events" : ["event1"]}}') }

      tracker = LastEventTracker.new(pathToFile, 15)

      # given multiple log events in the different streams
      log1 = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
      log1.timestamp = 1
      log1.log_stream_name = 'streamX'
      log1.event_id = 'event1'

      log2 = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
      log2.timestamp = log1.timestamp
      log2.log_stream_name = 'streamY'
      log2.event_id = 'event2'
      
      log3 = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()      
      log3.timestamp = log1.timestamp
      log3.log_stream_name = 'streamZ'
      log3.event_id = 'event3'

      tracker.check_event_already_processed('groupA', log1)
      tracker.check_event_already_processed('groupA', log2)
      tracker.check_event_already_processed('groupB', log3)

      # and the model is saved
      tracker.save()
      
      # and we create a new tracker and reload the data from the saved file
      tracker = LastEventTracker.new(pathToFile, 15)
      tracker.load()

      expect(tracker.get('groupA', log1).events).to contain_exactly(log1.event_id)
      expect(tracker.get('groupA', log1).timestamp).to eq(1)                 

      expect(tracker.get('groupA', log2).events).to contain_exactly(log2.event_id)
      expect(tracker.get('groupA', log2).timestamp).to eq(1)        

      expect(tracker.get('groupB', log3).events).to contain_exactly(log3.event_id)
      expect(tracker.get('groupB', log3).timestamp).to eq(1)        
    end

    it "test file handling on first run"   do

      # given a tracker to a file that doesn't exist
      tracker = LastEventTracker.new("not a real path", 15)

      # load it
      tracker.load()

      # given a new log event
      log = Aws::CloudWatchLogs::Types::FilteredLogEvent.new()
      log.message = 'this be the verse'
      log.timestamp = 1
      log.log_stream_name = 'streamX'
      log.event_id = 'event1'

      # then the log event should be there
      expect(tracker.get('group', log)).to be_falsey
      # and it should work
      tracker.check_event_already_processed('group', log)
      expect(tracker.get('group', log).events).to contain_exactly(log.event_id)

    end

  end

end


ci/setup.sh
#!/bin/bash
# version: 1
########################################################
#
# AUTOMATICALLY GENERATED! DO NOT EDIT
#
########################################################
set -e
if [ "$LOGSTASH_BRANCH" ]; then
    echo "Building plugin using Logstash source"
    BASE_DIR=`pwd`
    echo "Checking out branch: $LOGSTASH_BRANCH"
    git clone -b $LOGSTASH_BRANCH https://github.com/elastic/logstash.git ../../logstash --depth 1
    printf "Checked out Logstash revision: %s\n" "$(git -C ../../logstash rev-parse HEAD)"
    cd ../../logstash
    echo "Building plugins with Logstash version:"
    cat versions.yml
    echo "---"
    # We need to build the jars for that specific version
    echo "Running gradle assemble in: `pwd`"
    ./gradlew assemble
    cd $BASE_DIR
    export LOGSTASH_SOURCE=1
else
    echo "Building plugin using released gems on rubygems"
fi

lib/logstash/inputs/last_event_tracker.rb
# encoding: utf-8
require "fileutils"
require 'json'
java_import org.apache.logging.log4j.LogManager

# Tracks the last events for a group
# Contains only the last timestamp, and the events for that timestamp
class LastEvents    

  def initialize(timestamp = 0, events = [])
    @timestamp = timestamp
    @events = events
  end

  def timestamp
    @timestamp
  end

  def events
    @events
  end

  # Check the timestamp / event id
  # Returns true: If the timestamp/event is already in the model
  # Returns false: if the timestamp is not new, but the event is
  # Returns false: if the timestamp is new. This will cause the entire model to be flushed
  def check_event_already_processed(timestamp, event_id)
    if timestamp == @timestamp 
      if @events.include? event_id
        return true
      end 
      @events.push(event_id)
      return false
    else
      @timestamp = timestamp
      @events.clear()
      @events.push(event_id)
      return false
    end
  end
 

end

# Tracks the last events for each log stream
#
# In all cases log_event is one of the events from 
# https://docs.aws.amazon.com/sdk-for-ruby/v2/api/Aws/CloudWatchLogs/Client.html#filter_log_events-instance_method
class LastEventTracker
  include LogStash::Util::Loggable

  def initialize(path_to_data, prune_since_db_stream_minutes)
    @path_to_data = path_to_data
    @last_events_by_stream = Hash.new
    @logger = LogManager.getLogger(LastEventTracker)
    @prune_since_db_stream_minutes = prune_since_db_stream_minutes
  end

  # Given a log group and log event
  # - Return truthy if event was already processed, or falsy, otherwise
  # - Will record all the event ids with for the last timestamp   
  def check_event_already_processed(log_group, log_event)
    id = identify(log_group, log_event)
    if !@last_events_by_stream.key?(id)
      @last_events_by_stream[id] = LastEvents.new
    end

    return @last_events_by_stream[id].check_event_already_processed(log_event.timestamp, log_event.event_id)
  end

  # Get the last event ids for the given group/event
  def get(log_group, log_event)
    id = identify(log_group, log_event)
    return @last_events_by_stream[id]
  end



  # saves the data model to json 
  def save() 
    # build the json model
    data = {}
    @last_events_by_stream.each do |k,v| 
      data[k] = {
        "timestamp" => @last_events_by_stream[k].timestamp,
        "events" => @last_events_by_stream[k].events
      }
    end  
    
    
    # save it
    begin      
      File.write(@path_to_data, data.to_json)
    rescue Errno::EACCES
      # probably no file handles free
      # maybe it will work next time
      @logger.debug? && @logger.debug("Failed to write to: #{@path_to_data}: #{$!}")
    end    
  end

  def load
    @logger.debug("Loading config from: #{@path_to_data}")  

    begin
      @last_events_by_stream.clear
      if File.file?(@path_to_data)        
        data_hash = JSON.parse(File.read(@path_to_data))
        data_hash.each do |k, v|
          @last_events_by_stream[k] = LastEvents.new(v['timestamp'], v['events'])
        end
      end
    rescue 
      # if we can't read the file, we just assume it's broken, and we'll ignore it
      @logger.debug("Failed to read: #{@path_to_data}: #{$!}")  
    end
  end 

  # Create a key to uniquely identify the group/stream for the log event
  private
  def identify(log_group, log_event)
    return log_group + '.' + log_event.log_stream_name
  end

  private 
  def purge_old()
    # find the max time
    max_time = 0
    @last_events_by_stream.each do |k, v|
      if max_time < v['timestamp']
        max_time = v['timestamp']
      end
    end

    # build the lookback time
    purge_before = (max_time - (60 * 1000 * @prune_since_db_stream_minutes))  

    # prune the old stream data
    @last_events_by_stream.clone.each do |k, v|
      if v['timestamp'] < purge_before
        @sincedb.delete(k)
      end
    end    
  end  
  
end 


lib/logstash/inputs/cloudwatch_logs.rb
# encoding: utf-8
require "logstash/inputs/base"
require "logstash/namespace"
require "logstash/plugin_mixins/aws_config"
require "logstash/timestamp"
require "time"
require "stud/interval"
require "aws-sdk"
require "logstash/inputs/cloudwatch_logs/patch"
require "fileutils"
require 'logstash/inputs/group_event_tracker'


Aws.eager_autoload!

# Stream events from CloudWatch Logs streams.
#
# Specify an individual log group, and this plugin will scan
# all log streams in that group, and pull in any new log events.
#
# Optionally, you may set the `log_group_prefix` parameter to true
# which will scan for all log groups matching the specified prefix
# and ingest all logs available in all of the matching groups.
#
class LogStash::Inputs::CloudWatch_Logs < LogStash::Inputs::Base
  include LogStash::PluginMixins::AwsConfig::V2

  config_name "cloudwatch_logs"

  default :codec, "plain"

  # Log group(s) to use as an input. If `log_group_prefix` is set
  # to `true`, then each member of the array is treated as a prefix
  config :log_group, :validate => :string, :list => true

  # Where to write the since database (keeps track of the date
  # the last handled log stream was updated). The default will write
  # sincedb files to some path matching "$HOME/.sincedb*"
  # Should be a path with filename not just a directory.
  config :sincedb_path, :validate => :string, :default => nil
  # the stream data grows over time, so we drop it after a configurable time
  # but only after a new value comes in for some group (i.e. we purge one group 
  # at a time)
  config :prune_since_db_stream_minutes, :validate => :number, :default => 60

  # Interval to wait between to check the file list again after a run is finished.
  # Value is in seconds.
  config :interval, :validate => :number, :default => 60

  # Decide if log_group is a prefix or an absolute name
  config :log_group_prefix, :validate => :boolean, :default => false

  # Decide if present, then the results of the log group query are filtered again
  # to limit to these values. Only applicable if log_group_prefix = true 
  config :log_group_suffix, :validate => :string, :list => true, :default => nil  
  config :negate_log_group_suffix, :validate => :boolean, :default => false

  # When a new log group is encountered at initial plugin start (not already in
  # sincedb), allow configuration to specify where to begin ingestion on this group.
  # Valid options are: `beginning`, `end`, or an integer, representing number of
  # seconds before now to read back from.
  config :start_position, :default => 'beginning'
  


  # def register
  public
  def register
    require "digest/md5"
    @logger.debug("Registering cloudwatch_logs input", :log_group => @log_group)
    settings = defined?(LogStash::SETTINGS) ? LogStash::SETTINGS : nil
    check_start_position_validity

    Aws::ConfigService::Client.new(aws_options_hash)
    @cloudwatch = Aws::CloudWatchLogs::Client.new(aws_options_hash)

    if @sincedb_path.nil?
      if settings
        datapath = File.join(settings.get_value("path.data"), "plugins", "inputs", "cloudwatch_logs")
        # Ensure that the filepath exists before writing, since it's deeply nested.
        FileUtils::mkdir_p datapath
        @sincedb_path = File.join(datapath, ".sincedb_" + Digest::MD5.hexdigest(@log_group.join(",")))
      end
    end

    # This section is going to be deprecated eventually, as path.data will be
    # the default, not an environment variable (SINCEDB_DIR or HOME)
    if @sincedb_path.nil? # If it is _still_ nil...
      if ENV["SINCEDB_DIR"].nil? && ENV["HOME"].nil?
        @logger.error("No SINCEDB_DIR or HOME environment variable set, I don't know where " \
                      "to keep track of the files I'm watching. Either set " \
                      "HOME or SINCEDB_DIR in your environment, or set sincedb_path in " \
                      "in your Logstash config for the file input with " \
                      "path '#{@path.inspect}'")
        raise
      end

      #pick SINCEDB_DIR if available, otherwise use HOME
      sincedb_dir = ENV["SINCEDB_DIR"] || ENV["HOME"]

      @sincedb_path = File.join(sincedb_dir, ".sincedb_" + Digest::MD5.hexdigest(@log_group.join(",")))

      @logger.info("No sincedb_path set, generating one based on the log_group setting",
                   :sincedb_path => @sincedb_path, :log_group => @log_group)      
    end
  
    @logger.info("Using sincedb_path #{@sincedb_path}")
    @event_tracker = LogEventTracker.new(@sincedb_path, @prune_since_db_stream_minutes)
  end #def register


  public
  def check_start_position_validity
    raise LogStash::ConfigurationError, "No start_position specified!" unless @start_position

    return if @start_position =~ /^(beginning|end)$/
    return if @start_position.is_a? Integer

    raise LogStash::ConfigurationError, "start_position '#{@start_position}' is invalid! Must be `beginning`, `end`, or an integer."
  end # def check_start_position_validity

  # def run
  public
  def run(queue)
    @queue = queue
    @priority = []
    @event_tracker.load()

    while !stop?
      begin
        groups = find_log_groups

        groups.each do |group|
          @logger.debug("calling process_group on #{group}")
          process_group(group)
        end # groups.each
      rescue Aws::CloudWatchLogs::Errors::ThrottlingException
        @logger.debug("reached rate limit")
      end

      Stud.stoppable_sleep(@interval) { stop? }
    end
  end # def run

  public
  def find_log_groups
    if @log_group_prefix
      @logger.debug("log_group prefix is enabled, searching for log groups")
      groups = []
      next_token = nil
      @log_group.each do |group|
        loop do
          log_groups = @cloudwatch.describe_log_groups(log_group_name_prefix: group, next_token: next_token)
          # if we have no suffix setting, or if the candidate group name ends with the suffix
          # we use it
          groups += log_groups.log_groups            
            .select { |n| @log_group_suffix.nil?  || (n.log_group_name.end_with?(*@log_group_suffix) ^ @negate_log_group_suffix)}
            .map {|n| n.log_group_name}
          
          next_token = log_groups.next_token
          @logger.debug("found #{log_groups.log_groups.length} log groups matching prefix #{group}")
          break if next_token.nil?
        end
      end
    else
      @logger.debug("log_group_prefix not enabled")
      groups = @log_group
    end
    # Move the most recent groups to the end
    groups.sort{|a,b| priority_of(a) <=> priority_of(b) }
  end # def find_log_groups

  private
  def priority_of(group)
    @priority.index(group) || -1
  end


  private
  def process_group(group)
    next_token = nil
    loop do
      start_time = @event_tracker.get_or_set_min_time(group, get_default_start_time)

      params = {
          :log_group_name => group,
          :start_time => start_time,
          :interleaved => true,
          :next_token => next_token
      }      
      resp = @cloudwatch.filter_log_events(params)
    
      actually_processed_count = 0
      resp.events.each do |event|
        was_processed = process_log(event, group)
        was_processed && actually_processed_count = actually_processed_count + 1
      end

      resp.events.length() > 0 &&  @logger.debug("Queried logs for #{group} from #{parse_time(start_time)} found #{resp.events.length()} events, processed #{actually_processed_count}")
      # prune old records before saving
      @event_tracker.purge(group)
      @event_tracker.save()

      next_token = resp.next_token
      break if next_token.nil?
    end
    @priority.delete(group)
    @priority << group

  end #def process_group

  # def process_log - returns true if the message was actually processed
  private
  def process_log(log, group)
    identity = identify(group, log.log_stream_name)        
    if @event_tracker.is_new_event(group, log)
      @logger.trace? && @logger.trace("Processing event")    
      @codec.decode(log.message.to_str) do |event|

        event.set("@timestamp", parse_time(log.timestamp))
        event.set("[cloudwatch_logs][ingestion_time]", parse_time(log.ingestion_time))
        event.set("[cloudwatch_logs][log_group]", group)
        event.set("[cloudwatch_logs][log_stream]", log.log_stream_name)
        event.set("[cloudwatch_logs][event_id]", log.event_id)
        decorate(event)

        @queue << event

        @event_tracker.record_processed_event(group, log)
        return true
      end      
    end
    return false
  end # def process_log

  # def parse_time
  private
  def parse_time(data)
    LogStash::Timestamp.at(data.to_i / 1000, (data.to_i % 1000) * 1000)
  end # def parse_time


  private 
  def identify(group, log_stream_name)
    # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-logs-loggroup.html
    # ':' isn't allowed in a log group name, so we can use it safely
    return "#{group}:#{log_stream_name}"
  end

  private
  def is_stream_identifier(sincedb_name) 
    return sincedb_name.include? ":"
  end

  private
  def get_default_start_time()
    # chose the start time based on the configs
    case @start_position
    when 'beginning'
      return 0
    when 'end'
      return DateTime.now.strftime('%Q').to_i
    else
      return DateTime.now.strftime('%Q').to_i - (@start_position.to_i * 1000)
    end # case @start_position    
  end


end # class LogStash::Inputs::CloudWatch_Logs
